src/error.rs

```diff
@@ -5,78 +5,166 @@
 // <LICENSE-MIT or https://opensource.org/licenses/MIT>, at your
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
-use core::convert::From;
 use core::fmt;
 use core::num::NonZeroU32;
 
-// A randomly-chosen 24-bit prefix for our codes
-pub(crate) const CODE_PREFIX: u32 = 0x57f4c500;
-const CODE_UNKNOWN: u32 = CODE_PREFIX | 0x00;
-const CODE_UNAVAILABLE: u32 = CODE_PREFIX | 0x01;
-
-/// The error type.
+/// A small and `no_std` compatible error type.
 ///
-/// This type is small and no-std compatible.
+/// The [`Error::raw_os_error()`] will indicate if the error is from the OS, and
+/// if so, which error code the OS gave the application. If such an error is
+/// encountered, please consult with your system documentation.
 #[derive(Copy, Clone, Eq, PartialEq)]
-pub struct Error(pub(crate) NonZeroU32);
+pub struct Error(NonZeroU32);
+
+// This NonZeroU32 in Error has enough room for two types of errors:
+//   - OS Errors:       in range [1, 1 << 31) (i.e. positive i32 values)
+//   - Custom Errors:   in range [1 << 31, 1 << 32) (in blocks of 1 << 16)
+const CUSTOM_START: u32 = 1 << 31;
+const BLOCK_SIZE: u32 = 1 << 16;
 
 impl Error {
-    /// An unknown error.
-    pub const UNKNOWN: Error = Error(unsafe { NonZeroU32::new_unchecked(CODE_UNKNOWN) });
+    /// Create a new error from a raw OS error number (errno).
+    pub fn from_os_error(errno: i32) -> Self {
+        assert!(errno > 0);
+        Self(NonZeroU32::new(errno as u32).unwrap())
+    }
 
-    /// No generator is available.
-    pub const UNAVAILABLE: Error = Error(unsafe { NonZeroU32::new_unchecked(CODE_UNAVAILABLE) });
+    /// Crate a custom error in the provided block (group of 2^16 error codes).
+    /// The provided block must not be negative, and block 0 is reserved for
+    /// custom errors in the `getrandom` crate.
+    pub fn custom_error(block: i16, code: u16) -> Self {
+        assert!(block >= 0);
+        let n = CUSTOM_START + (block as u32) * BLOCK_SIZE + (code as u32);
+        Self(NonZeroU32::new(n).unwrap())
+    }
 
-    /// Extract the error code.
+    /// Extract the raw OS error code (if this error came from the OS)
     ///
-    /// This may equal one of the codes defined in this library or may be a
-    /// system error code.
+    /// This method is identical to `std::io::Error::raw_os_error()`, except
+    /// that it works in `no_std` contexts. If this method returns `None`, the
+    /// error value can still be formatted via the `Diplay` implementation.
+    pub fn raw_os_error(&self) -> Option<i32> {
+        self.try_os_error().ok()
+    }
+
+    /// Extract the bare error code.
     ///
-    /// One may attempt to format this error via the `Display` implementation.
+    /// This code can either come from the underlying OS, or be a custom error.
+    /// Use [`raw_os_error()`] to disambiguate.
     pub fn code(&self) -> NonZeroU32 {
```

Comment:
Worth adding `#[inline]` here and e.g. to `raw_os_error`? (Without it, inlining is not possible across crate boundaries.)

Reply:
Fixed. Added `#[inline]` to all of the `pub` methods, as we want to inline the bit-twiddling across crate boundaries.